extensions/git/src/model.ts

```diff
@@ -543,6 +543,14 @@ export class Model implements IRemoteSourcePublisherRegistry, IPostCommitCommand
 					}
 				}
 
+				// If this path is not the root folder of this repo
+				// and it contains it's own .git folder, then consider it a separate repo
+				if (!pathEquals(liveRepository.repository.root, resourcePath)) {
+					if (fs.existsSync(path.join(resourcePath, '.git'))) {
```

Comment:
Sure, that seems like a better idea, let me whip something up..

Reply:
Actually this might be better.

`git rev-parse --show-toplevel`

You might need to see what the minimum version of git that VSCode supports is

Reply:
Right.. this gets more complicated because `getRepository -> getOpenRepository` is synchronous. This is what is called from `openRepository` and ultimately the cause of the problem. So, there is no straight forward way to insert a `this.git.exec` directly into `getOpenRepository`. 

An alternate way to solve this might be to replace the `this.getRepository(repoPath)` check inside `openRepository`? This would also fix the problem, but then `getOpenRepository `is still arguably not implemented correctly. It will continue to report subdirectories that contain their own repositories as belonging to the outer repository, which is the root of this problem.

What are your thoughts about this?

Reply:
> Actually this might be better.
> 
> `git rev-parse --show-toplevel`
> 
> You might need to see what the minimum version of git that VSCode supports is

Yes, in fact, it looks like `Git.getRepositoryRoot` does exactly this already. It would be ideal to compare the repository root from this call for the subdirectory to the root of the outer repository. Alas, the async problem. 

Maybe I can try to write a specialized synchronous version of Git.exec for this specific problem..

Or maybe we don't care that getOpenRepository is wrong? Honestly, I was a bit surprised getOpenRepository works the way it does, I thought it would _only_ match the root directory for the repo. I'm not entirely sure why it's like this or what the intent was.

Reply:
Sorry for the flurry of responses. I think I've convinced myself it's better to simply change `openRepository`. I think the solution is call `this.git.getRepositoryRoot` for the `repoPath` in `openRepository`, then use this root in the call to `getRepository`.

Without really understanding the intent behind why `getRepository` is implemented the way it is, it's probably best not to mess around with it. The original change I made has this same problem, it might have further reaching unintended consequences.

I'll test the new changes and push them up shortly

Reply:
Unfortunately, this idea does not work. Even though the nested repositories are being "opened", the source control tab does not display them. Strangely, the blue circle notification on the source control tab _does_ show the correct number of files that have been edited, as though the nested repositories are opened.

I assume `getOpenRepository` is called elsewhere that results in the same problem, so it looks like the fix really needs to be in this function.

Reply:
I also briefly looked at trying to create a Git.getRepositoryRootSync function and this is a much bigger job that I'm willing to commit to. There is heavy usage of async functionality everywhere in this class. I tried to hack together something really quickly as a helper function inside Model, and that was also a minefield, lots of path specific transforms and such that need to happen. The code will be very fragile without putting in the Git class to ensure all the logic matches up correctly.

Which brings us back to checking for the .git folder directly. If this is simply unacceptable, I can just close the pull request. Unless you have other suggestions, it's clear to me how to proceed on this.

Reply:
It isn't clear to me why the function can't be async. I am sure someone from VSCode will have some good ideas.

Reply:
`Model.getOpenRepository` has two callsites that aren't async. One might be possible to change, but the other is `Model.getRepository`, which is exported and has more callsites that aren't async. At a cursory glance, it doesn't look straight forward to simply propagate async _until it works_. The effects will be cascading.

Reply:
@nteymory, @tristan957, thank you for keeping the conversation going. As I was looking at the code, I pretty much went down the same path as the two of you, and I have to admit that at the moment I am missing some historical context around the implementation of `getOpenRepository()` and `getRepository()`. I am hoping to get that context next week and then I will circle back to this pull request.

Reply:
@lszomoru is there anything I can do to move this pull request forward? I would love to see this problem fixed.

Reply:
This week, the team was heads down preparing the next stable release. Having said that, I did look at the code and I have created a new pull request (https://github.com/microsoft/vscode/pull/159291) that adds a new setting as well as some new code to support the new setting while keeping the existing code path in case the setting is disabled. I will try to get this pull request in next week but if you want please feel free to check out the branch, build, and test the new setting. Thank you!

Reply:
It looks like your changes work. I will close this pull request